gnd/src/main/java/com/google/android/gnd/ui/home/HomeScreenViewModel.java

```diff
@@ -85,8 +85,10 @@
                         .toSingleDefault(newFeature)
                         .doOnError(this::onAddFeatureError)
                         .onErrorResumeNext(Single.never())) // Prevent from breaking upstream.
+            .filter(feature -> feature.getLayer().getForm().isPresent())
+            .doOnNext(this::addNewObservation)
```

Comment:
Relying on side-effects to do actual work is an anti-pattern that's we're actively seeking to avoid; rather than using `filter` + `doOnNext` can we just add a method called from `subscribe`, called `onAddFeature` which checks if the form is present and if so calls, `addNewObservation`?

Reply:
Thanks for the wisdom!

Reply:
My pleasure! Added https://github.com/google/ground-android/wiki/RxJava-Usage-and-Conventions#side-effects to clarify.